### PR TITLE
Issue #663 - option to disable de-serialization using companion constructors

### PR DESCRIPTION
**Pull request by BlueTufa**

### CODE REVIEW

**Comment by magnolia-k** [comment]
@xuwei-k 
Please review.

File: README.md [hunk 1]
```diff
@@ -705,6 +705,13 @@ Both these settings (`strictOptionParsing` and `strictArrayExtraction`) can be e
 val formats: Formats = DefaultFormats.strict
 ```
 
+With Json4s 3.6 and higher, `apply` functions in companion objects will be evaluated for use during extraction.  If this behavior is not desired, you can disable it using the `alwaysConsiderCompanionConstructors` on a custom `Formats` object:
+```scala 
+val formats: Formats = new DefaultFormats { override val alwaysConsiderCompanionConstructors = false }
+```
+
+When this option is disbled, only primary and secondary constructors will be evaluated for use during extraction.
```
Comment (mccartney) [inline README.md:713]:
```suggestion
When this option is disabled, only primary and secondary constructors will be evaluated for use during extraction.
```

Comment (mccartney) [reply README.md:713]:
```suggestion
When this option is disabled, only primary and secondary constructors will be evaluated for use during extraction.
```

File: core/src/main/scala/org/json4s/reflect/Reflector.scala [hunk 2]
```diff
@@ -42,18 +42,26 @@ object Reflector {
...
   def describe[T](implicit mf: Manifest[T], formats: Formats = DefaultFormats): ObjectDescriptor =
-    describe(scalaTypeDescribable(scalaTypeOf[T])(formats))
+    describe2(scalaTypeDescribable(scalaTypeOf[T]))
 
+  @deprecated("Use describe[T] or describe2 instead", "3.6.8")
   def describe(st: ReflectorDescribable[_]): ObjectDescriptor =
-    descriptors(st.scalaType, createDescriptor(_, st.paranamer, st.companionClasses))
+    descriptors(st.scalaType, createDescriptor2(_, st.paranamer, st.companionClasses)(DefaultFormats))
 
+  def describe2(st: ReflectorDescribable[_])(implicit formats: Formats): ObjectDescriptor =
```
Comment (mccartney) [inline core/src/main/scala/org/json4s/reflect/Reflector.scala:51]: Can we find a better name for this `describe2`?
I know the reason for having a different name then `describe`, but this `2` looks ugly.

File: core/src/main/scala/org/json4s/Formats.scala [hunk 3]
```diff
@@ -96,6 +96,14 @@ trait Formats extends Serializable { self: Formats =>
   def alwaysEscapeUnicode: Boolean = false
   def strictFieldDeserialization: Boolean = false
 
+  /**
+   * Setting to false preserves library's behavior prior to 3.6, where companion object constructors were only
+   * considered when deserializing if there were no primary constructors. Setting to true preserves the
+   * backwards-incompatible change made in 3.6 to always consider companion object constructors when deserializing
+   * (https://github.com/json4s/json4s/pull/487).
+   */
+  def alwaysConsiderCompanionConstructors: Boolean = true
```
Comment (mccartney) [inline core/src/main/scala/org/json4s/Formats.scala:105]: What does the `always` refer to here?
Shouldn't the setting be called `considerCompanionConstructors`?

File: core/src/main/scala/org/json4s/reflect/Reflector.scala [hunk 4]
```diff
@@ -124,16 +132,21 @@ object Reflector {
...
         }
       }
+
       val constructorDescriptors = createConstructorDescriptors(ccs)
-      companion = findCompanion(checkCompanionMapping = false)
-      val applyMethods: scala.Array[Method] = companion match {
-        case Some(singletonDescriptor) =>
-          singletonDescriptor.instance.getClass.getMethods.filter { method => method.getName == "apply" && method.getReturnType == er }
-        case None => scala.Array[Method]()
+      val all = {
```
Comment (BlueTufa) [inline core/src/main/scala/org/json4s/reflect/Reflector.scala:137]: Would you prefer that this be refactored into a separate function?

Comment (mccartney) [reply core/src/main/scala/org/json4s/reflect/Reflector.scala:137]: I think it's fine for now.

**Pull request merged**